submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,313 @@
+
+    // Observed status of GlobalnetEgressIP. Its a read-only field.
+    // +optional
+    Status             EgressIPStatus `json:"status,omitempty"`
+}
+
+type EgressIPSpec struct {
+    // GlobalIPs is a list of EgressIPs requested. The requested globalIPs should be part
+    // of globalCIDR of the Cluster. If the requested globalIPs are not allocated to other
+    // K8s objects, then it will be allocated and the status field will be updated accordingly.
+    // If the globalIPs are already allocated to other objects, creation of the CRD will fail.
+    // If this field is not specified, a single globalIP will be allocated.
+    // This field cannot be changed through updates.
+    // +optional
+    GlobalIPs  []string `json:"globalIPs,omitempty"`
+
+    // Selects the Namespaces to which this GlobalnetEgressIP object applies.
+    // If an empty namespaceSelector: {} is configured, it selects all the Pods in all namespaces.
+    // On the other hand, if you omit specifying namespaceSelector, it does not select any
+    // namespaces and selects only Pods from the namespace where the GlobalnetEgressIP is
+    // deployed to.
+    // +optional
+    NamespaceSelector metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+
+    // Selects the pods whose label match the definition. This is an optional field and
+    // in case it is not set, it results in all the Pods selected from the NamespaceSelector.
+    // In case it is set, its intersected with NamespaceSelector, and only Pods that match the
+    // PodSelector and present in the NamespaceSelector will have the GlobalIPs as EgressIPs.
+    // Either namespaceSelector or podSelector have to be specified. If both are omitted,
+    // its considered as an error.
+    // When a Pod or Namespace matches multiple GlobalnetEgressIP objects, there is no guarantee
+    // which of the globalIP address that are assigned to GlobalnetEgressIP will be used.
+    // +optional
+    PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
+}
+
+const (
+    // GlobalnetEgressIPSuccess means that Globalnet was able to successfully
+    // allocate the GlobalIP as requested in GlobalnetEgressIP object.
+    GlobalnetEgressIPSuccess GlobalnetEgressIPStatus = "Success"
+
+    // GlobalnetEgressIPError means that Globalnet was unable to allocate the
+    // GlobalIP for the requested GlobalnetEgressIP object.
+    GlobalnetEgressIPError GlobalnetEgressIPStatus = "Error"
+)
+
+type EgressIPStatus struct {
+    // Status is one of {"Success", "Error"}
+    Status GlobalnetEgressIPStatus `json:"status,omitempty"`
+
+    // +optional
+    Message *string `json:"message,omitempty"`
+
+    // The list of GlobalIPs assigned via this GlobalnetEgressIP object.
+    GlobalIPs []string `json:"globalIPs"`
+
+    // The Namespaces to which the GlobalIPs are applied.
+    Namespaces []string `json:"namespaces"`
+
+    // The Pods to which the GlobalIPs are applied.
+    Pods []string `json:"pods"`
```

Comment:
While probably easier to store this information here, I feel that if we annotate the namespaces/pods that get a global egress ip with the active policy(or policies) that affect them it'll be better for the users as they'll more easily see which policy affects their pods/namespaces.

Does this make sense?

Reply:
Right. I want to take some time to decide whether we want to annotate each and every Pod/Namespace as it requires an update operation. As its an implementation detail, we can decide how we want to proceed later.